File: test-start-workflow/fixed/vulnerable.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

// Fix 1: Safe strcpy with bounds checking
void safe_strcpy(char *dest, const char *src, size_t dest_size) {
    strncpy(dest, src, dest_size - 1);
    dest[dest_size - 1] = '\0';
}

// Fix 2: Removed use after free
void safe_memory_demo() {
    char *ptr = malloc(100);
    strcpy(ptr, "This is safe");
    printf("%s\n", ptr);
    free(ptr);
    // No access after free
}

// Fix 3: Fixed memory leak
char* create_buffer_safe() {
    char *buffer = malloc(256);
    // Caller must free this buffer
    return buffer;
}

// Fix 4: Integer overflow check
void safe_allocation(int size) {
    if (size > 0 && size < INT_MAX / sizeof(char)) {
        char *buffer = malloc(size * sizeof(char));
        if (buffer) {
            // Use buffer
            free(buffer);
        }
    }
}

// Fix 5: Safe format string
void print_user_input_safe(char *input) {
    printf("%s", input);  // Safe format string
}

int main(int argc, char *argv[]) {
    char small_buffer[10];
    
    if (argc > 1) {
        safe_strcpy(small_buffer, argv[1], sizeof(small_buffer));
        print_user_input_safe(argv[1]);
    }
    
    safe_memory_demo();
    char *buffer = create_buffer_safe();
    free(buffer);  // Properly freed
    safe_allocation(1000000000);
    
    printf("Program completed safely\n");
    return 0;
}

File: test-start-workflow/vulnerable.rs
use std::env;

// Rust equivalent - Memory safe by default!

fn safe_copy(dest: &mut [u8], src: &[u8]) {
    let len = dest.len().min(src.len());
    dest[..len].copy_from_slice(&src[..len]);
}

fn safe_memory_demo() {
    let data = String::from("This is safe");
    println!("{}", data);
    // Memory automatically freed (RAII)
}

fn create_buffer_safe() -> Vec<u8> {
    vec![0; 256]
    // Returns owned Vec, memory managed safely
}

fn safe_allocation(size: usize) {
    if size > 0 {
        let buffer = vec![0u8; size];
        // Automatically freed at end of scope
    }
}

fn print_user_input_safe(input: &str) {
    println!("{}", input);  // Format string safety by default
}

fn main() {
    let args: Vec<String> = env::args().collect();
    
    if args.len() > 1 {
        let mut small_buffer = [0u8; 10];
        safe_copy(&mut small_buffer, args[1].as_bytes());
        print_user_input_safe(&args[1]);
    }
    
    safe_memory_demo();
    let _buffer = create_buffer_safe();
    safe_allocation(1000000000);
    
    println!("Program completed safely");
}
